File: weekly-contest-132/recover-a-tree-from-preorder-traversal/recover-a-tree-from-preorder-traversal.js
function TreeNode(val) {
    this.val = val
    this.left = this.right = null
}

const recoverFromPreorder = str => {
    const reg = /(\-*\d+)/g

    const parse_token = token => {
        let depth = 0
        const val = []
        for (const x of token) {
            if (x == '-') {
                depth += 1
                continue
            }
            val.push(x)
        }
        return { depth, val: +val.join('') }
    }

    const stack = []
    const h = tokens => {
        if (!tokens.length) {
            return
        }

        const token = tokens[0]
        const { depth, val } = parse_token(token)
        const node = new TreeNode(val)

        while (stack[stack.length - 1].depth >= depth) {
            stack.pop()
        }
        const { node: parent } = stack[stack.length - 1]
        if (parent.left) {
            parent.right = node
        } else {
            parent.left = node
        }
        stack.push({ depth, node })

        h(tokens.slice(1))
    }

    const tokens = str.match(reg)
    const root = new TreeNode(+tokens[0])
    stack.push({ node: root, depth: 0 })
    h(tokens.slice(1))

    return root
}

const serialize = root => {
    const r = []
    const queue = [root]
    while (queue.length) {
        const node = queue.shift()
        if (node) {
            r.push(node.val)
            queue.push(node.left)
            queue.push(node.right)
            continue
        }
        r.push('null')
    }

    console.log(r.join(', '))
}
serialize(recoverFromPreorder('1-2--3---4-5--6---7'))

File: weekly-contest-132/recover-a-tree-from-preorder-traversal/recover-a-tree-from-preorder-traversal.cpp
#include <algorithm>
#include <functional>
#include <iostream>
#include <iterator>
#include <map>
#include <queue>
#include <set>
#include <sstream>
#include <stack>
#include <string>
#include <vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

// "1-2--3--4-5--6--7"
class Solution {
  public:
    TreeNode *recoverFromPreorder(string s) {
        vector<tuple<int, int>> tokens;
        auto val = -1;
        auto depth = 0;
        for (auto x : s) {
            if (x == '-') {
                if (val != -1) {
                    tokens.push_back({depth, val});
                    val = -1;
                    depth = 0;
                }
                depth += 1;
                continue;
            }
            if (val != -1) {
                val = val * 10 + (x - '0');
                continue;
            }
            val = x - '0';
        }
        tokens.push_back({depth, val});

        stack<tuple<int, TreeNode *>> stack;
        function<void(int)> h = [&](auto i) {
            if (i >= tokens.size()) {
                return;
            }
            auto [depth, val] = tokens[i];

            while (get<0>(stack.top()) >= depth) {
                stack.pop();
            }
            auto node = new TreeNode(val);
            auto parent = get<1>(stack.top());
            if (parent->left) {
                parent->right = node;
            } else {
                parent->left = node;
            }

            stack.push({depth, node});
            h(i + 1);
        };

        auto root = new TreeNode(get<1>(tokens[0]));
        stack.push({0, root});
        h(1);

        return root;
    }
};

int main() {
    auto s = new Solution();
    auto a = "1-2--3--4-5--6--7";
    auto r = s->recoverFromPreorder("3");
}